src/client_builder.rs

```diff
@@ -61,41 +155,132 @@ where
         ClientBuilder {
             domain,
             port,
-            starttls: false,
+            mode: ConnectionMode::Auto,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            tls_kind: TlsKind::Any,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            no_tls_verify: false,
         }
     }
 
     /// Use [`STARTTLS`](https://tools.ietf.org/html/rfc2595) for this connection.
     #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    #[deprecated = "Use mode(ConnectionMode::StartTls) instead"]
     pub fn starttls(&mut self) -> &mut Self {
-        self.starttls = true;
+        self.mode(ConnectionMode::StartTls)
+    }
+
+    /// Use TLS for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    #[deprecated = "Use mode(ConnectionMode::Tls) instead"]
+    pub fn tls(&mut self) -> &mut Self {
+        self.mode(ConnectionMode::Tls)
+    }
+
+    /// Sets the Connection mode to use for this connection
+    pub fn mode(&mut self, mode: ConnectionMode) -> &mut Self {
+        self.mode = mode;
         self
     }
 
-    /// Return a new [`Client`] using a `native-tls` transport.
+    /// Sets the TLS backend to use for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
+        self.tls_kind = kind;
+        self
+    }
+
+    /// Controls the use of certificate validation.
+    ///
+    /// Defaults to `false`.
+    ///
+    /// # Warning
+    ///
+    /// You should only use this as a last resort as it allows another server to impersonate the
+    /// server you think you're talking to, which would include being able to receive your
+    /// credentials.
+    ///
+    /// See [`native_tls::TlsConnectorBuilder::danger_accept_invalid_certs`],
+    /// [`native_tls::TlsConnectorBuilder::danger_accept_invalid_hostnames`],
+    /// [`rustls::ClientConfig::dangerous`]
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn danger_no_tls_verify(&mut self, no_tls_verify: bool) -> &mut Self {
+        self.no_tls_verify = no_tls_verify;
```

Comment:
The double negation here strikes me as unfortunate. Can we name the field and the method just `danger_tls_verify` and `tls_verify`, and have the latter default to `true` instead?

Reply:
Alternatively, `danger_skip_tls_verify`.

Reply:
Skip tls verify sounds better. as the default should be to verify w/o needing to call anything.